crates/relayer-types/src/core/ics04_channel/timeout.rs

```diff
@@ -204,31 +206,73 @@ pub enum UpgradeTimeout {
     Both(Height, Timestamp),
 }
 
-impl UpgradeTimeout {
+impl Display for Timeout {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        match self {
+            Self::Height(height) => write!(f, "{height}"),
+            Self::Timestamp(timestamp) => write!(f, "{timestamp}"),
+            Self::Both(height, timestamp) => write!(f, "{height}, {timestamp}"),
+        }
+    }
+}
+
+impl Timeout {
     pub fn new(height: Option<Height>, timestamp: Option<Timestamp>) -> Result<Self, ChannelError> {
         match (height, timestamp) {
-            (Some(height), None) => Ok(UpgradeTimeout::Height(height)),
-            (None, Some(timestamp)) => Ok(UpgradeTimeout::Timestamp(timestamp)),
-            (Some(height), Some(timestamp)) => Ok(UpgradeTimeout::Both(height, timestamp)),
+            (Some(height), None) => Ok(Timeout::Height(height)),
+            (None, Some(timestamp)) => Ok(Timeout::Timestamp(timestamp)),
+            (Some(height), Some(timestamp)) => Ok(Timeout::Both(height, timestamp)),
             (None, None) => Err(ChannelError::missing_upgrade_timeout()),
         }
     }
 
     pub fn into_tuple(self) -> (Option<Height>, Option<Timestamp>) {
         match self {
-            UpgradeTimeout::Height(height) => (Some(height), None),
-            UpgradeTimeout::Timestamp(timestamp) => (None, Some(timestamp)),
-            UpgradeTimeout::Both(height, timestamp) => (Some(height), Some(timestamp)),
+            Timeout::Height(height) => (Some(height), None),
+            Timeout::Timestamp(timestamp) => (None, Some(timestamp)),
+            Timeout::Both(height, timestamp) => (Some(height), Some(timestamp)),
+        }
+    }
+}
+
+define_error! {
+    #[derive(Debug, PartialEq, Eq)]
+    TimeoutError {
+        InvalidTimestamp
+            { timestamp: String }
+            [ TraceError<ParseTimestampError> ]
+            |e| { format_args!("cannot convert into a `Timestamp` type from string {0}", e.timestamp) },
+
+        InvalidTimeout
+            { timeout: String }
+            |e| { format_args!("invalid timeout {0}", e.timeout) },
+    }
+}
+
+impl FromStr for Timeout {
+    type Err = TimeoutError;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        let split: Vec<&str> = value.split(' ').collect();
```

Comment:
I believe the format of the string will be `{<number>-<height>} <timestamp>`, but need to double check, so that's why I split on the empty space.